### PR TITLE
3.0 - Build script support (image build engine)

### PR DESCRIPTION
**Pull request by bauerm97**
**Description of the Pull Request (PR):**

Updated and cleaned up version of #1668. 

- Rebased onto upstream
- Cleaned up git commit history 
- Removed code referencing C SIF and Image libs
- Fix issues regarding permissions of calling user
- Prevent SUID escalation for imgbuild engine

### CODE REVIEW

File: src/cmd/singularity/cli/sif.go [hunk 1]
```diff
@@ -28,8 +28,8 @@ func init() {
 	SifCmd.Flags().SetInterspersed(false)
 	SingularityCmd.AddCommand(SifCmd)
 	// Create
-	SifCmd.AddCommand(SifCreate)
```
Comment (yhcote) [inline src/cmd/singularity/cli/sif.go]: I think we can graciously remove anything related to the create command here. I agree that it should not be there anymore. There should not be two different path from singularity to create a SIF image and "build" is the right command to accomplish that.

File: src/cmd/singularity/cli/sif.go [hunk 2]
```diff
@@ -55,10 +55,25 @@ var SifCmd = &cobra.Command{
...
+		}
+		if err := sif.Wait(); err != nil {
+			sylog.Fatalf("sif failed: %v\n", err)
+		}
+
+	},
+	DisableFlagParsing: true,
 }
 
-// SifCreate sif create cmd
```
Comment (yhcote) [inline src/cmd/singularity/cli/sif.go]: This too should be completely removed as opposed to commented :)

**Review by yhcote — DISMISSED** [review]
Nothing fancy but removing unused lines are satisfying ;)

I would also make sure there's no more references to "singularity sif create" from comments or docs for the cli.

File: mlocal/frags/Makefile.stub [hunk 3]
```diff
@@ -67,7 +67,7 @@ $(wrapper_OBJ): $(libstartup) $(SOURCEDIR)/src/runtime/startup/c/wrapper.c
 $(wrapper): $(wrapper_OBJ) $(libstartup)
 	@echo " LD" $@
 	@mkdir -p $(@D)
-	$(V)$(CC) $(LDFLAGS) -o $@ $(wrapper_OBJ) -Wl,--start-group -L$(BUILDDIR)/lib -lruntime -lstartup -Wl,--end-group -luuid -lpthread $(WRAPPER_LIBS)
+	$(V)$(CC) $(LDFLAGS) -o $@ $(wrapper_OBJ) -Wl,--start-group -L$(BUILDDIR)/lib -lruntime -lstartup -Wl,--end-group -luuid -lpthread -lgpgme $(WRAPPER_LIBS)
```
Comment (yhcote) [inline mlocal/frags/Makefile.stub:70]: What is the gpgme for ? Do we import a package that requires it ?

Comment (bauerm97) [reply mlocal/frags/Makefile.stub:70]: Yes, this is an unfortunate side effect of using `build.Definition` as a member field of `src/runtime/engines/imgbuild/config.EngineConfig`. `build` package needs `-lgpgme`. Maybe this means the `Definition` type should be moved to a buildtypes subpackage?

File: src/runtime/engines/singularity/create.go [hunk 4]
```diff
@@ -98,118 +187,127 @@ func (engine *EngineOperations) CreateContainer(pid int, rpcConn net.Conn) error
...
-		return fmt.Errorf("mount /dev failed: %s", err)
+	// rootfs is an image
+	return mountImage(rpcOps, mnt)
+}
+
+// mount any generic mount (not loop dev)
+func mountGeneric(rpcOps *client.RPC, mnt specs.Mount) error {
+	flags, opts, _ := mount.ConvertOptions(mnt.Options)
+	optsString := ""
+	for _, opt := range opts {
```
Comment (cclerget) [inline src/runtime/engines/singularity/create.go]: you can use strings.Join here ;)

File: src/runtime/engines/singularity/create.go [hunk 5]
```diff
@@ -98,118 +187,127 @@ func (engine *EngineOperations) CreateContainer(pid int, rpcConn net.Conn) error
...
+	_, err := rpcOps.Mount(mnt.Source, mnt.Destination, mnt.Type, flags, optsString)
+	return err
+}
+
+// mount image via loop
+func mountImage(rpcOps *client.RPC, mnt specs.Mount) error {
+	flags, opts, iopts := mount.ConvertOptions(mnt.Options)
+
+	optsString := ""
+	for _, opt := range opts {
```
Comment (cclerget) [inline src/runtime/engines/singularity/create.go]: here too

**Review by cclerget — CHANGES_REQUESTED** [review]
Looks good Michael ! Just suggested 2 minor changes

**Review by cclerget — APPROVED** [review]

**Pull request merged**